app/models/miq_server/worker_management/monitor/kubernetes.rb

```diff
@@ -53,45 +96,68 @@ def orchestrator
     @orchestrator ||= ContainerOrchestrator.new
   end
 
+  def monitor_deployments
+    loop do
+      current_deployments.clear
+      resource_version = collect_initial(:deployments)
+
+      watch_for_events(:deployments, resource_version)
+    end
+  end
+
   def monitor_pods
     loop do
       current_pods.clear
-      resource_version = collect_initial_pods
+      resource_version = collect_initial(:pods)
 
-      # watch_for_pod_events doesn't return unless an error caused us to break out of it, so we'll start over again
-      watch_for_pod_events(resource_version)
+      # watch_for_events doesn't return unless an error caused us to break out of it, so we'll start over again
+      watch_for_events(:pods, resource_version)
     end
   end
 
-  def collect_initial_pods
-    pods = orchestrator.get_pods
-    pods.each { |p| save_pod(p) }
-    pods.resourceVersion
+  def collect_initial(resource = :pods)
+    objects = orchestrator.send("get_#{resource}")
+    objects.each { |p| send("save_#{resource.to_s.singularize}", p) }
+    objects.resourceVersion
   end
+  # TODO: Remove callers to collect_initial_pods
+  alias_method :collect_initial_pods, :collect_initial
 
-  def watch_for_pod_events(resource_version)
-    orchestrator.watch_pods(resource_version).each do |event|
+  def watch_for_events(resource, resource_version)
+    orchestrator.send("watch_#{resource}", resource_version).each do |event|
       case event.type.downcase
       when "added", "modified"
-        save_pod(event.object)
+        send("save_#{resource.to_s.singularize}", event.object)
       when "deleted"
-        delete_pod(event.object)
+        send("delete_#{resource.to_s.singularize}", event.object)
       when "error"
         if (status = event.object)
           # ocp 3 appears to return 'ERROR' watch events with the object containing the 410 code and "Gone" reason like below:
           # #<Kubeclient::Resource type="ERROR", object={:kind=>"Status", :apiVersion=>"v1", :metadata=>{}, :status=>"Failure", :message=>"too old resource version: 199900 (27177196)", :reason=>"Gone", :code=>410}>
-          log_pod_error_event(status.code, status.message, status.reason)
+          log_resource_error_event(status.code, status.message, status.reason)
         end
 
         break
       end
     end
   end
+  # TODO: Remove callers of watch_for_events
+  alias_method :watch_for_pod_events, :watch_for_events
 
-  def log_pod_error_event(code, message, reason)
+  def log_resource_error_event(code, message, reason)
     _log.warn("Restarting watch_for_pod_events due to error: [#{code} #{reason}], [#{message}]")
   end
 
+  def save_deployment(deployment)
+    name = deployment.metadata.name
+    current_deployments[name] ||= Concurrent::Hash.new
+    current_deployments[name].merge!(deployment.to_h)
```

Comment:
Is this storing the entire deployment?

Reply:
Yes.  We need `[:spec][:replicas]` for `scale` and `[:spec][:template][:spec]`... for resource constraints.  It's not even an easy lookup for the latter since it's so deeply nested so I can't even pluck just the parts I want that easily.

```ruby
    container = current_deployments.fetch_path(worker.worker_deployment_name, :spec, :template, :spec, :containers).try(:first)
    current_constraints = container.try(:fetch, :resources, nil) || {}
```

I could save the toplevel `:spec` and not the rest but I don't think the remaining parts are that large and changing the  `save` would make the `query` part need match that change.  If this works, we can trim away things if it's too large.  I was thinking that we would not have that many deployments we're caching, maybe less than 10.

Reply:
I did some checking on what's cached in the current_deployments.  We can save a bit of memory by only saving `[:spec]` and below:

```
irb(main):005:0> s.current_deployments.to_s.length
=> 27950
irb(main):006:0> s.current_deployments.keys.length
=> 8

irb(main):007:0> s.current_deployments['1-generic']
=> {:metadata=>{:name=>"1-generic", :namespace=>"jrafanie-test-patch-deployment-two", :selfLink=>"/apis/apps/v1/namespaces/jrafanie-test-patch-deployment-two/deployments/1-generic", :uid=>"b302c53c-38c7-11eb-9834-d094660d31fb", :resourceVersion=>"52956101", :generation=>4, :creationTimestamp=>"2020-12-07T20:06:31Z", :labels=>{:app=>"manageiq", :"manageiq-orchestrated-by"=>"orchestrator-689cf6c4b-qcmq5"}, :annotations=>{:"deployment.kubernetes.io/revision"=>"3"}, :ownerReferences=>[{:apiVersion=>"v1", :kind=>"Pod", :name=>"orchestrator-689cf6c4b-qcmq5", :uid=>"c77f2e37-38c6-11eb-9834-d094660d31fb", :controller=>true, :blockOwnerDeletion=>true}]}, :spec=>{:replicas=>2, :selector=>{:matchLabels=>{:name=>"1-generic"}}, :template=>{:metadata=>{:name=>"1-generic", :creationTimestamp=>nil, :labels=>{:app=>"manageiq", :"manageiq-orchestrated-by"=>"orchestrator-689cf6c4b-qcmq5", :name=>"1-generic"}}, :spec=>{:containers=>[{:name=>"1-generic", :image=>"jrafanie/manageiq-base-worker:patch-deploy3", :env=>[{:name=>"DATABASE_PORT", :value=>"5432"}, {:name=>"GUID", :value=>"be6379f3-38c6-11eb-9834-d094660d31fb"}, {:name=>"MEMCACHED_SERVER"}, {:name=>"MEMCACHED_SERVICE_NAME"}, {:name=>"WORKER_HEARTBEAT_FILE", :value=>"/var/www/miq/vmdb/tmp/worker.hb"}, {:name=>"WORKER_HEARTBEAT_METHOD", :value=>"file"}, {:name=>"DATABASE_HOSTNAME", :valueFrom=>{:secretKeyRef=>{:name=>"postgresql-secrets", :key=>"hostname"}}}, {:name=>"DATABASE_NAME", :valueFrom=>{:secretKeyRef=>{:name=>"postgresql-secrets", :key=>"dbname"}}}, {:name=>"DATABASE_PASSWORD", :valueFrom=>{:secretKeyRef=>{:name=>"postgresql-secrets", :key=>"password"}}}, {:name=>"DATABASE_USER", :valueFrom=>{:secretKeyRef=>{:name=>"postgresql-secrets", :key=>"username"}}}, {:name=>"ENCRYPTION_KEY", :valueFrom=>{:secretKeyRef=>{:name=>"app-secrets", :key=>"encryption-key"}}}, {:name=>"WORKER_CLASS_NAME", :value=>"MiqGenericWorker"}, {:name=>"BUNDLER_GROUPS", :value=>"manageiq_default,ui_dependencies"}], :resources=>{:limits=>{:cpu=>"990m", :memory=>"600Mi"}, :requests=>{:cpu=>"150m", :memory=>"499Mi"}}, :livenessProbe=>{:exec=>{:command=>["/usr/local/bin/manageiq_liveness_check"]}, :initialDelaySeconds=>240, :timeoutSeconds=>10, :periodSeconds=>15, :successThreshold=>1, :failureThreshold=>3}, :terminationMessagePath=>"/dev/termination-log", :terminationMessagePolicy=>"File", :imagePullPolicy=>"Always"}], :restartPolicy=>"Always", :terminationGracePeriodSeconds=>600, :dnsPolicy=>"ClusterFirst", :serviceAccountName=>"manageiq-default", :serviceAccount=>"manageiq-default", :securityContext=>{}, :schedulerName=>"default-scheduler"}}, :strategy=>{:type=>"RollingUpdate", :rollingUpdate=>{:maxUnavailable=>"25%", :maxSurge=>"25%"}}, :revisionHistoryLimit=>10, :progressDeadlineSeconds=>600}, :status=>{:observedGeneration=>4, :replicas=>3, :updatedReplicas=>1, :readyReplicas=>2, :availableReplicas=>2, :unavailableReplicas=>1, :conditions=>[{:type=>"Available", :status=>"True", :lastUpdateTime=>"2020-12-07T20:08:27Z", :lastTransitionTime=>"2020-12-07T20:08:27Z", :reason=>"MinimumReplicasAvailable", :message=>"Deployment has minimum availability."}, {:type=>"Progressing", :status=>"True", :lastUpdateTime=>"2020-12-07T20:16:51Z", :lastTransitionTime=>"2020-12-07T20:06:31Z", :reason=>"ReplicaSetUpdated", :message=>"ReplicaSet \"1-generic-5bbb7bcc55\" is progressing."}]}}

irb(main):008:0> s.current_deployments['1-generic'].to_s.length
=> 3343
irb(main):010:0> s.current_deployments['1-generic'][:spec].to_s.length
=> 2098
```

Reply:
~63% of the cache is the `:spec` key and below.  We can save about ~37% by only getting that.

Reply:
I've implemented this in the commit "Save the :spec subkey, discarding others like metadata/status."